### ISSUE TITLE
Implement first iteration of `MockChainContext`

### ISSUE THREAD

**Issue by soareschen** [informalsystems/hermes#2733]
## Summary

The new relayer is fully generic and can work with any concrete chain implementation. With that we can create a simplified mock chain to test the relayer logic.

The mock chain can be implemented by implementing the OneForAll traits and instantiated with naive types such as:

```rust
struct PacketKey {
  client_id: u128,
  channel_id: u128,
  port_id: u128,
  sequence: u128,
  ...
}

struct MockChain {
  packet_commitments: HashMap<PacketKey, String>,
}

impl OfaChainTypes for MockChain {
  type Height = u128;
  type ClientId = u128;
  ...
}
```

Using this, we can test the correctness of the relayer by inspecting the end state of the mock chain, and see whether it records the correct state.

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate milestone (priority) applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by soareschen** [comment]
@ljoss17 there are some follow up fixes that we still need to work on after #2785 is merged:

- The semantics of `process_messages` should be atomic: the messages need to be either all being processed successfully at the same height, or all failed. So this means that methods like `self.receive_packet` should not directly update the chain state. Instead they should take the current pending chain state, and return a new chain state to be committed. The next message handler would take the chain state returned from the previous message handler. And if any message handling failed, the whole pending chain state is discarded.

- The `query_write_ack_event` method introduced by #2907 is currently stubbed with `todo!()`, and needs to be implemented.

- The `Message::RecvPacket`, `Message::AckPacket`, and `Message::TimeoutPacket` should not contain the client ID. The client ID should be retrieved by having a mapping from channel ID to client ID in the local chain state.

- The `PacketKey` used in `Message` should contain the following fields: `source_channel_id`, `source_port_id`, `destination_channel_id`, `destination_port_id`, `sequence`, `timeout_height`, `timeout_timestamp`. Whether to read from the source or channel ID depends on the specific message. For example, processing of `RecvPacket` would use the destination channel ID as the local channel ID.

**Comment by seanchen1991** [comment]
To be able to track this more cleanly, I opened #2957 to keep track of the follow-up work.
